lib/models/room.js

```diff
@@ -237,43 +316,226 @@ Room.prototype.getAvatarUrl = function(baseUrl, width, height, resizeMethod,
+        throw Error("timeline support is disabled. Set the 'timelineSupport'" +
+                    " parameter to true when creating MatrixClient to enable" +
+                    " it.");
+    }
+
+    var timeline = new EventTimeline(this.roomId);
+    this._timelines.push(timeline);
+    return timeline;
+};
+
+
+/**
+ * Add events to a timeline
+ *
+ * <p>Will fire "Room.timeline" for each event added.
+ *
  * @param {MatrixEvent[]} events A list of events to add.
- * @param {boolean} toStartOfTimeline True to add these events to the start
+ *
+ * @param {boolean} toStartOfTimeline   True to add these events to the start
  * (oldest) instead of the end (newest) of the timeline. If true, the oldest
  * event will be the <b>last</b> element of 'events'.
+ *
+ * @param {module:models/event-timeline~EventTimeline=} timeline   timeline to
+ *    add events to. If not given, events will be added to the live timeline
+ *
+ * @param {string=} paginationToken   token for the next batch of events
+ *
  * @fires module:client~MatrixClient#event:"Room.timeline"
+ *
  */
-Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline) {
-    var stateContext = toStartOfTimeline ? this.oldState : this.currentState;
+Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline,
+                                              timeline, paginationToken) {
+    if (!timeline) {
+        timeline = this._liveTimeline;
+    }
 
-    function checkForRedaction(redactEvent) {
-        return function(e) {
-            return e.getId() === redactEvent.event.redacts;
-        };
+    if (!toStartOfTimeline && timeline == this._liveTimeline) {
+        // special treatment for live events
+        this._addLiveEvents(events);
+        return;
+    }
+
+    // we need to handle the following case:
+    //
+    // We already know about two timelines. One of these contains just
+    // event M, and one contains just event P:
+    //
+    //    timeline1      timeline2
+    //       [M]            [P]
+    //
+    // We are now processing a new pagination response from the server.
+    // This contains the following events: [M, N, P, R, S].
+    //
+    // 1. First, we ignore event M, since we already know about it.
+    //
+    // 2. Next, we append N to timeline 1.
+    //
+    // 3. Next, we don't add event P, since we already know about it,
+    //    but we do link togehter the timelines. We now have:
+    //
+    //    timeline1      timeline2
+    //     [M, N] <----> [P]
+    //
+    // 4. We now need to append R and S to timeline2. So finally, we have:
+    //
+    //    timeline1      timeline2
+    //     [M, N] <----> [P, R, S]
+    //
+    // The important thing to note in the above is that we switched the
+    // timeline we were adding events to, after we found a neighbour.
+
+    // XXX: TODO: but timeline2 might already be joined to timeline3: we need
+    // to skip all the way to the end of the linked list!
```

Comment:
Wonderful prose, thank you! As for the `XXX TODO`: This is presumably unavoidable that we need to follow the linked list and scan each element in each list? The comment suggests that there may be a way to avoid doing this? It might be worth stating your proposal even if you don't implement it now.